File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
import sys
sys.setrecursionlimit(10000)
def solution(maps):
    from collections import deque

    def dfs(start_x, start_y, graph,cnt):
        
        if maps[start_x][start_y] == 'X' :
            return
        cnt += int(maps[start_x][start_y])
        visited.add((start_x,start_y))

        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]

        for i in range(4):
            nx = (start_x) + dx[i]
            ny = (start_y) + dy[i]

            if 0 <= nx < len(graph) and 0 <= ny < len(graph[0]):
                if (nx, ny) not in visited and maps[nx][ny] != 'X':
                    visited.add((nx, ny)) 
                    cnt = dfs(nx,ny,maps,cnt)
                          
        return cnt
    
    visited = set()
    result = []
    
    for i in range(len(maps)):
        for j in range(len(maps[0])):
            if (i, j) not in visited and maps[i][j] != 'X':
                result.append(dfs(i,j,maps,0))
    if not result :
        return [-1]
    else :
        return sorted(result)

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.util.*;
import java.io.*;
class Solution {
    /*
    상하좌우 갈 수 있으면 끝까지 간다.
    연결시켜서 최대 합을 더함..
    그 후 오름차순 정렬
    
    알고리즘 분류 : DFS? 영역 분리니까..
    어떻게 풀이? MAPS 전체를 돌면서 X가 아니면 visited에 있는지 체크하고 X아닌 걸 상하좌우 다 넣음..
    
    */
    
    
    public int[] solution(String[] maps) {
        int n = maps.length;
        int m = maps[0].length();
        boolean[][] visited = new boolean[n][m];
        List<Integer> islandFoods = new ArrayList<>();
        
        
        
        
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if(maps[i].charAt(j) != 'X' && !visited[i][j]){
                    int totalFood = dfs(maps,visited,i,j);
                    islandFoods.add(totalFood);
                }
            }
        }
        
        if (islandFoods.isEmpty()) {
            return new int[] {-1};
        }
        Collections.sort(islandFoods);
        
        int[] answer = new int[islandFoods.size()];
        for (int i = 0; i < islandFoods.size(); i++) {
            answer[i] = islandFoods.get(i);
        }
        return answer;
        
    }
    private int dfs(String[] maps, boolean[][] visited, int i, int j){
        int n = maps.length;
        int m = maps[0].length();
        
        int[] dx = {-1,1,0,0};
        int[] dy = {0,0,-1,1};
        int cnt = 0;
        if (i>=0 && i< n && j>=0 && j < m  && !visited[i][j] && maps[i].charAt(j) != 'X'){
            visited[i][j] = true;
            
            cnt = maps[i].charAt(j) - '0';
            for(int k=0; k<4; k++){
                int nx = i + dx[k];
                int ny = j + dy[k];
                cnt+= dfs(maps,visited,nx,ny);
            }
        }
        return cnt;
        
        
    }
}